service/replication.go

```diff
@@ -475,131 +469,92 @@ func (s *service) GetStorageProtectionGroupStatus(ctx context.Context, req *csie
-	s1TP, err := isiConfig.isiSvc.client.GetTargetPolicyByName(ctx, ppName)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "can't find remote replication policy, unexpected error %s", err.Error())
-	}
-
-	// obtain target policy for target cluster
-	s2P, err := remoteIsiConfig.isiSvc.client.GetPolicyByName(ctx, ppName)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "can't find remote replication policy, unexpected error %s", err.Error())
-	}
-
-	var state csiext.StorageProtectionGroupStatus_State
-	if (s1P.Enabled && !s2P.Enabled && s1TP.FailoverFailbackState == "writes_enabled" && s2TP.FailoverFailbackState == "writes_disabled") ||
-		(!s1P.Enabled && s2P.Enabled && s1TP.FailoverFailbackState == "writes_disabled" && s2TP.FailoverFailbackState == "writes_enabled") {
-		state = csiext.StorageProtectionGroupStatus_SYNCHRONIZED
-	} else if localSusp && ((s1P.Enabled || !s1P.Enabled) && !s2P.Enabled && s1TP.FailoverFailbackState == "writes_disabled" && (s2TP == nil || s2TP.FailoverFailbackState == "writes_enabled")) ||
-		(!s1P.Enabled && (s2P.Enabled || !s2P.Enabled) && (s1TP == nil || s1TP.FailoverFailbackState == "writes_enabled") && s2TP.FailoverFailbackState == "writes_disabled") {
-		state = csiext.StorageProtectionGroupStatus_SUSPENDED
-	} else if ((s1P.Enabled || !s1P.Enabled) && !s2P.Enabled && s1TP.FailoverFailbackState == "writes_disabled" && (s2TP == nil || s2TP.FailoverFailbackState == "writes_enabled")) ||
-		(!s1P.Enabled && (s2P.Enabled || !s2P.Enabled) && (s1TP == nil || s1TP.FailoverFailbackState == "writes_enabled") && s2TP.FailoverFailbackState == "writes_disabled") {
-		state = csiext.StorageProtectionGroupStatus_FAILEDOVER
-	} else if isSync {
-		state = csiext.StorageProtectionGroupStatus_SYNC_IN_PROGRESS
-	} else if s1TP.LastJobState == "failed" || s2TP.LastJobState == "failed" {
-		state = csiext.StorageProtectionGroupStatus_INVALID
-	} else {
-		state = csiext.StorageProtectionGroupStatus_UNKNOWN
+	if linkState == csiext.StorageProtectionGroupStatus_UNKNOWN || isSyncCheckFailed {
+		errMsg := "unexpected error while getting link state"
+		if isSyncCheckFailed {
+			errMsg = "unexpected error while querying active jobs for local or remote policy"
+		}
+		log.Error(errMsg)
+		resp := &csiext.GetStorageProtectionGroupStatusResponse{
+			Status: &csiext.StorageProtectionGroupStatus{
+				State: csiext.StorageProtectionGroupStatus_UNKNOWN,
+				// do not update isSource
+			},
+		}
+		return resp, status.Errorf(codes.Internal, errMsg)
 	}
 
-	log.Info("trying to get replication direction")
+	log.Info("Trying to get replication direction")
 	source := false
-	if s1P.Enabled {
+	if linkState != csiext.StorageProtectionGroupStatus_FAILEDOVER && // no side can be source when in failed over state
+		(localP.Enabled || // when synchronized
+			(!remoteP.Enabled && localTP.FailoverFailbackState == "writes_enabled" && remoteTP.FailoverFailbackState == "writes_disabled")) { // when suspended (source side)
```

Comment:
Minor cleanup issue, these two strings are used in a lot of places. May be worth consolidating into a set of constants.